amethyst_network/src/simulation/resource.rs

```diff
@@ -0,0 +1,192 @@
+        Self {
+            socket: None,
+            server_addr: None,
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: true,
+        }
+    }
+
+    /// Add a number of trusted clients to the `NetworkSimulationResource` for use as a server
+    pub fn with_trusted_clients(mut self, clients: &[SocketAddr]) -> Self {
+        self.clients = clients.iter().map(|addr| Client::new(*addr)).collect();
+        self
+    }
+
+    /// Returns a slice of the tracked clients
+    pub fn clients(&self) -> &[Client] {
+        &self.clients
+    }
+
+    /// Set the server address
+    pub fn set_server_addr(&mut self, server_addr: Option<SocketAddr>) {
+        self.server_addr = server_addr;
+    }
+
+    /// Set the bound socket to the `NetworkSimulationResource`
+    pub fn set_socket(&mut self, socket: S) {
+        self.socket = Some(socket);
+    }
+
+    /// Drops the socket from the `NetworkSimulationResource`
+    pub fn drop_socket(&mut self) {
+        self.socket = None;
+    }
+
+    /// Returns whether or not the `NetworkSimulationResource` has a bound socket
+    pub fn has_socket(&self) -> bool {
+        self.socket.is_some()
+    }
+
+    /// Returns the server address if one was set
+    pub fn server_addr(&self) -> Option<SocketAddr> {
+        self.server_addr
+    }
+
+    /// Returns whether or not this `NetworkSimulationResource` was created as a 'server'
+    pub fn is_server(&self) -> bool {
+        self.is_server
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent on next sim tick.
+    pub fn send(&mut self, payload: &[u8]) {
+        self.send_with_requirements(payload, S::default_guarantee(), UrgencyRequirement::OnTick);
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent immediately.
+    pub fn send_immediate(&mut self, payload: &[u8]) {
+        self.send_with_requirements(
+            payload,
+            S::default_guarantee(),
+            UrgencyRequirement::Immediate,
+        );
+    }
+
+    /// Create and queue a `Message` with the specified guarantee
+    pub fn send_with_requirements(
+        &mut self,
+        payload: &[u8],
+        delivery: DeliveryRequirement,
+        timing: UrgencyRequirement,
+    ) {
+        let message = Message::new(payload, delivery, timing);
+        self.messages.push_back(message);
+    }
+
+    /// Drains the messages queue and returns the drained messages
+    pub fn drain_messages(&mut self, mut filter: impl FnMut(&mut Message) -> bool) -> Vec<Message> {
+        let mut drained = Vec::with_capacity(self.messages.len());
+        let mut i = 0;
+        while i != self.messages.len() {
+            if filter(&mut self.messages[i]) {
+                let message = self.messages.remove(i);
```

Comment:
`self.messages.remove(i)`
Ouch, that may be painful performance-wise.
Until `drain_filter` is available for `VecDeque` and stabilized, let's do it at least in two iterations with `.iter().filter(filter)` and `.retain(filter)`? Or what about introducing a real separate queue for immediate messages? That would let us avoid redundant iterations for `OnTick` messages each frame at all.

p.s. but thanks for implementing `UrgencyRequirement` anyway ;)

Reply:
Yeah that's probably fine. Would be pretty straight forward to implement and is likely faster